File: app/src/main/java/com/example/ogan/codedenim/Courses/ModuleActivity.java
package com.example.ogan.codedenim.Courses;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;

import com.example.ogan.codedenim.Adapters.ModuleAdapter;
import com.example.ogan.codedenim.Gson.CourseGson.Module;
import com.example.ogan.codedenim.Gson.Module.GetModule;
import com.example.ogan.codedenim.Gson.Module.ModuleApus;
import com.example.ogan.codedenim.R;

import java.util.ArrayList;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ModuleActivity extends AppCompatActivity {

    private static final String url = "https://codedenim.azurewebsites.net/api/";
    ModuleAdapter moduleAdapter;
    GetModule getModule;
    RecyclerView recyclerView;
    LinearLayoutManager linearLayoutManager;
    ArrayList<ModuleApus> results;
    ProgressBar progressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_module);

        recyclerView = (RecyclerView) findViewById(R.id.module_RV);
        progressBar = (ProgressBar) findViewById(R.id.module_pr);
        linearLayoutManager = new LinearLayoutManager(getApplicationContext(), LinearLayoutManager.HORIZONTAL, false);
        recyclerView.setLayoutManager(linearLayoutManager);

        Intent intent = getIntent();
        String courseName = intent.getStringExtra("courseName");
        int courseId = intent.getIntExtra("courseId", 0);

        if(getSupportActionBar() != null ){
            getSupportActionBar().setTitle(courseName + " Modules");
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        Retrofit retrofit = new Retrofit.Builder().baseUrl(url).
                addConverterFactory(GsonConverterFactory.create()).build();

        getModule = retrofit.create(GetModule.class);


        //GET modules
        Call<ArrayList<ModuleApus>> call = getModule.getModuleList(courseId);
        call.enqueue(new Callback<ArrayList<ModuleApus>>() {
            @Override
            public void onResponse(Call<ArrayList<ModuleApus>> call, Response<ArrayList<ModuleApus>> response) {


                System.out.println(response.message());
                if(response.isSuccessful()){

                    results = response.body();
                    moduleAdapter = new ModuleAdapter(results);
                    recyclerView.setAdapter(moduleAdapter);
                    progressBar.setVisibility(View.GONE);
                    recyclerView.setVisibility(View.VISIBLE);

                }
            }

            @Override
            public void onFailure(Call<ArrayList<ModuleApus>> call, Throwable t) {

                Log.e("error", t.getMessage());
            }
        });

    }
}

File: app/src/main/java/com/example/ogan/codedenim/courses/ModuleActivity.kt
package com.example.ogan.codedenim.courses

import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.MenuItem
import android.view.View
import android.widget.ProgressBar
import com.example.ogan.codedenim.MyUtilClass
import com.example.ogan.codedenim.R
import com.example.ogan.codedenim.ServiceGenerator
import com.example.ogan.codedenim.adapters.ModuleAdapter
import com.example.ogan.codedenim.gson.Module
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class ModuleActivity : AppCompatActivity() {

    private var moduleAdapter: ModuleAdapter? = null
    private lateinit var recyclerView: RecyclerView
    private var results: ArrayList<Module>? = null
    private lateinit var progressBar: ProgressBar
    private var courseId: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.general_recyclerview_layout)

        recyclerView = findViewById(R.id.general_recyclerView)
        progressBar = findViewById(R.id.general_pr)
        val linearLayoutManager = LinearLayoutManager(applicationContext, LinearLayoutManager.VERTICAL, false)
        recyclerView.layoutManager = linearLayoutManager

        val intent = intent
        val courseName = intent.getStringExtra("courseName")
        courseId = intent.getIntExtra("courseId", 0)

        if (supportActionBar != null) {
            supportActionBar?.title = courseName + " Modules"
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }

        getData()

        val refreshLayout = findViewById<SwipeRefreshLayout>(R.id.swipe_refresh)
        refreshLayout.setOnRefreshListener {
            moduleAdapter = null
            progressBar.visibility = View.VISIBLE
            getData()
            refreshLayout.isRefreshing = false
        }

    }

    private fun getData() {
        if (MyUtilClass.checkNetworkConnection(applicationContext)) {
            //GET modules
            ServiceGenerator.apiMethods.getModuleList(courseId).enqueue(object : Callback<ArrayList<Module>> {
                override fun onResponse(call: Call<ArrayList<Module>>, response: Response<ArrayList<Module>>?) {

                    if (response != null) {
                        println(response.message())
                        if (response.isSuccessful) {

                            results = response.body()
                            println(response.body())
                            moduleAdapter = ModuleAdapter(results)
                            recyclerView.adapter = moduleAdapter
                            progressBar.visibility = View.GONE
                            recyclerView.visibility = View.VISIBLE

                        } else MyUtilClass.showAlert(this@ModuleActivity, "Sorry, an error occurred. Swipe down to refresh")
                    }
                }

                override fun onFailure(call: Call<ArrayList<Module>>, t: Throwable?) {

                    progressBar.visibility = View.GONE
                    MyUtilClass.showAlert(this@ModuleActivity, "Sorry, an error occurred. Swipe down to refresh")
                }
            })
        } else {
            progressBar.visibility = View.GONE
            MyUtilClass.showNoInternetMessage(this@ModuleActivity)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
        // Respond to the action bar's Up/Home button
            android.R.id.home -> {
                super@ModuleActivity.onBackPressed()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }
}
